wasm/src/validator.rs

```diff
@@ -0,0 +1,400 @@
+                            >::clone(&permission_token)
+                        )
+                    {
+                        return <$token as ::iroha_wasm::validator::traits::Validate>::validate_grant(
+                            &concrete_token,
+                            &$authority
+                        );
+                    }
+                )+}
+                if let $crate::data_model::prelude::Instruction::Revoke(revoke) = &$instruction {
+                    if let $crate::data_model::prelude::Value::PermissionToken(permission_token) =
+                        <
+                            $crate::data_model::prelude::EvaluatesTo<$crate::data_model::prelude::Value>
+                            as
+                            $crate::EvaluateOnHost
+                        >::evaluate_on_host(&revoke.object)
+                        .dbg_expect("Failed to evaluate `Revoke` object")
+                    {$(
+                        if let Ok(concrete_token) =
+                            <$token as ::core::convert::TryFrom<_>>::try_from(
+                                <
+                                    $crate::data_model::permission::token::Token as ::core::clone::Clone
+                                >::clone(&permission_token)
+                            )
+                        {
+                            return <$token as ::iroha_wasm::validator::traits::Validate>::validate_revoke(
+                                &concrete_token,
+                                &$authority
+                            );
+                        }
+                    )+}
+                }
+            }
+        };
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use webassembly_test::webassembly_test;
+
+        /// Can't be a doc-test because of `compile_error!` on native target and
+        /// `webassembly-test-runner` on wasm target.
+        #[webassembly_test]
+        fn test_deny() {
+            use crate::{
+                alloc::borrow::ToOwned as _, data_model::permission::validator::Verdict, deny,
+            };
+
+            let a = || deny!("Some reason");
+            assert_eq!(a(), Verdict::Deny("Some reason".to_owned()));
+
+            let get_reason = || "Reason from expression".to_owned();
+            let b = || deny!(get_reason());
+            assert_eq!(b(), Verdict::Deny("Reason from expression".to_owned()));
+
+            let mes = "Format message";
+            let c = || deny!("Reason: {}", mes);
+            assert_eq!(c(), Verdict::Deny("Reason: Format message".to_owned()));
+
+            let mes = "Advanced format message";
+            let d = || deny!("Reason: {mes}");
+            assert_eq!(
+                d(),
+                Verdict::Deny("Reason: Advanced format message".to_owned())
+            );
+        }
+    }
+}
+
+/// Error type for `TryFrom<PermissionToken>` implementations.
+///
+/// TODO: Use some `no_std` error library?
```

Comment:
There can't be such a library, because `std::error::Error` is not portable. Not a matter of importing `alloc`, I'm afraid. 

`derive_more::From` is the way to go, given that we won't be printing errors in a `no_std` environment in most cases.

Reply:
Usage of this error is generated by the `Token` derive macro and won't be used on user side, I believe. Do we really need `derive_more::From` then?

Reply:
I wanted you to remove or change this comment because it's incorrect. Whether or not deriving `From` makes sense is up to you